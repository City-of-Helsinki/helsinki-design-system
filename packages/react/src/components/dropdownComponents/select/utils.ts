import React from 'react';

import { SelectData, SelectMetaData, SelectDataHandlers, FilterFunction } from './types';
import {
  Group,
  Option,
  ModularOptionListData,
  ModularOptionListProps,
  ScreenReaderNotification,
} from '../modularOptionList/types';
import { ChangeEvent } from '../../dataProvider/DataContext';
import { eventTypes } from './events';
import {
  iterateAndCopyGroup,
  getGroupLabelOption,
  getAllOptions,
  propsToGroups,
  getSelectedOptions,
} from '../modularOptionList/utils';

// Re-export commonly used utility functions
export {
  getAllOptions,
  getSelectedOptions,
  propsToGroups,
  updateGroupLabelAndOptions,
  updateOptionInGroup,
  defaultFilter,
} from '../modularOptionList/utils';

type DomHandlerProps = {
  id: string;
  type?: string;
  trigger: (event: ChangeEvent) => void;
};

export type OptionIterator = (
  option: Option,
  group: Group,
  optionIndex: number,
  groupIndex: number,
) => Option | undefined;

export const DROPDOWN_MENU_ITEM_HEIGHT = 52;

export function createOnClickListener(props: DomHandlerProps) {
  const { id, type = eventTypes.click, trigger } = props;
  return {
    onClick: (originalEvent: React.MouseEvent) => {
      trigger({ id, type, payload: { originalEvent } });
    },
  };
}

export function createInputOnChangeListener(props: DomHandlerProps) {
  const { id, type = eventTypes.change, trigger } = props;
  return {
    onChange: (originalEvent: React.ChangeEvent<HTMLInputElement>) => {
      trigger({
        id,
        type,
        payload: { value: originalEvent.currentTarget.value, originalEvent },
      });
    },
  };
}

export function getElementIds(containerId: string): SelectMetaData['elementIds'] {
  return {
    container: containerId,
    button: `${containerId}-main-button`,
    list: `${containerId}-list`,
    clearButton: `${containerId}-clear-button`,
    label: `${containerId}-label`,
    selectionsAndListsContainer: `${containerId}-sl-container`,
    tagList: `${containerId}-tag-list`,
    searchOrFilterInput: `${containerId}-input-element`,
    searchOrFilterInputLabel: `${containerId}-input-element-label`,
    clearAllButton: `${containerId}-clear-all-button`,
    showAllButton: `${containerId}-show-all-button`,
  };
}

export function countVisibleOptions(groups: SelectData['groups']): number {
  let count = 0;
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (!option.isGroupLabel && option.visible) {
        count += 1;
      }
    });
  });
  return count;
}

export function createScreenReaderNotification(type: string, content: string, delay = 0): ScreenReaderNotification {
  return {
    type,
    content,
    delay,
    showTime: 0,
    addTime: Date.now(),
  };
}

export function filterOptions(groups: ModularOptionListData['groups'], filterStr: string, filterFunc: FilterFunction) {
  const newGroupsWithFilteredOptions = iterateAndCopyGroup(groups, (option) => {
    if (option.isGroupLabel) {
      return { ...option };
    }
    return { ...option, visible: !filterStr || filterFunc(option, filterStr) };
  });

  // check if group label should be visible....
  newGroupsWithFilteredOptions.forEach((group) => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    const groupLabel = getGroupLabelOption(group);
    if (!groupLabel) {
      return;
    }
    groupLabel.visible = !!groupLabel.label && group.options.findIndex((opt) => !opt.isGroupLabel && opt.visible) > -1;
  });

  return newGroupsWithFilteredOptions;
}

export function filterSelectableOptions(
  groups: ModularOptionListData['groups'],
  filterStr: string,
  filterFunc: FilterFunction,
  isMultiSelect: boolean,
) {
  return getAllOptions(groups, !isMultiSelect).filter((option) => filterFunc(option, filterStr));
}

export function mergeSearchResultsToCurrent(
  props: Pick<ModularOptionListProps, 'groups' | 'options'>,
  currentGroups: ModularOptionListData['groups'],
): ModularOptionListData['groups'] {
  const newData = propsToGroups(props) || [];
  const newOptions = getAllOptions(newData);
  const currentOptionsWithoutMatches = getSelectedOptions(currentGroups).filter((option) => {
    const sameInNewOptionsIndex = newOptions.findIndex((newOption) => {
      return newOption.value === option.value;
    });
    if (sameInNewOptionsIndex > -1) {
      newOptions[sameInNewOptionsIndex].selected = true;
      return false;
    }
    return true;
  });

  const currentHiddenOptionsInAGroup = currentOptionsWithoutMatches.length
    ? [{ options: currentOptionsWithoutMatches.map((opt) => ({ ...opt, visible: false })) } as Group]
    : [];

  return [...currentHiddenOptionsInAGroup, ...newData];
}

/**
 *
 * @param notification
 * @param dataHandlers
 * @returns Returns true, if notification was added
 */
export function addOrUpdateScreenReaderNotificationByType(
  notification: ScreenReaderNotification,
  dataHandlers: SelectDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfSameType = screenReaderNotifications.findIndex((n) => n.type === notification.type);
  if (indexOfSameType > -1) {
    const updatedList = [...screenReaderNotifications];
    const prev = updatedList[indexOfSameType];
    if (prev.content === notification.content) {
      return false;
    }
    updatedList[indexOfSameType] = notification;
    dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
    return false;
  }
  const updatedList = [...screenReaderNotifications, notification];
  dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
  return true;
}

export function removeScreenReaderNotification(
  target: Partial<ScreenReaderNotification>,
  dataHandlers: SelectDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfMatch = screenReaderNotifications.findIndex((n) => {
    const hasTypeMatch = !target.type || n.type === target.type;
    const hasContentMatch = !target.content || n.content === target.content;
    return hasTypeMatch && hasContentMatch;
  });
  if (indexOfMatch > -1) {
    screenReaderNotifications.splice(indexOfMatch, 1);
    dataHandlers.updateMetaData({ screenReaderNotifications });
    return true;
  }
  return false;
}
