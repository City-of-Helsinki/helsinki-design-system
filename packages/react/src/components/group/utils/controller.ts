import { Controller, ControllerProps } from '.';
import { createPropHandler } from './propHandler';
import { createStorage, StorageData } from './storage';

/**
 *
 * Controller for all Group's data and events.
 *
 */

export function createController({
  reRenderer,
  propSetter,
  initialData,
  onChange,
  metaData,
}: ControllerProps): Controller {
  const storage = createStorage(initialData);
  const meta = createStorage(metaData || {});
  const propHandler = createPropHandler(propSetter);
  return {
    getProps(props) {
      return propHandler.getProps({ ...props, controller: this });
    },
    getData() {
      return storage.get();
    },
    updateData({ data, updateKeys }) {
      storage.set(data);
      if (updateKeys && updateKeys.length) {
        updateKeys.forEach((id) => {
          propHandler.updateKey(id);
        });
      }
      reRenderer();
    },
    triggerChange(props) {
      return onChange({ ...props, controller: this });
    },
    getMetaData() {
      return meta.get() as StorageData;
    },
    updateMetaData(newData: StorageData) {
      return meta.set(newData);
    },
  };
}
