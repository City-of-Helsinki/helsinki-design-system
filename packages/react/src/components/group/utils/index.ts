import {
  SyntheticEvent,
  FunctionComponent,
  HtmlHTMLAttributes,
  PropsWithChildren,
  ComponentType,
  MutableRefObject,
} from 'react';

import { StorageData, Storage } from './storage';

export type ChangeEventPayload = { value?: unknown; originalEvent?: SyntheticEvent };
export type ChangeEvent = { id: string; type?: string; payload?: ChangeEventPayload };
export type ChangeHandlerProps = ChangeEvent & { controller: Controller };
export type ChangeHandler = (props: ChangeHandlerProps) => unknown;
export type HTMLElementAttributesWithChildren = PropsWithChildren<HtmlHTMLAttributes<HTMLElement>>;
export type DefaultGroupElementProps = Record<string, unknown> & {
  key?: string;
  forwardedController?: Controller;
  'data-hds-group-id'?: string;
};

export type ElementRef = MutableRefObject<HTMLElement | null>;

export type Controller = {
  getProps: ({ id: string, elementProps: DefaultGroupElementProps }) => ReturnType<PropSetter>;
  updateData: (props: { data: StorageData; updateKeys?: string[] }) => void;
  getData: Storage['get'];
  getMetaData: () => StorageData;
  updateMetaData: (appended: StorageData) => StorageData;
  triggerChange: (change: ChangeEvent) => unknown;
};

export type ControllerProps = {
  reRenderer: () => void;
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange: ChangeHandler;
};

export type PropSetter<T = DefaultGroupElementProps> = (props: {
  id: string;
  controller: Controller;
  key: string;
  elementProps: DefaultGroupElementProps;
}) => T;

export type PropHandler = {
  getProps: (props: {
    id: string;
    controller: Controller;
    elementProps: DefaultGroupElementProps;
  }) => DefaultGroupElementProps;
  updateKey: (id: string) => void;
};

export type GroupChildFunction = FunctionComponent<{ controller: Controller }>;

export type GroupChild<P = DefaultGroupElementProps> = ComponentType<HTMLElementAttributesWithChildren & P>;

export type GroupProps = PropsWithChildren<unknown> & {
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange?: ChangeHandler;
};
