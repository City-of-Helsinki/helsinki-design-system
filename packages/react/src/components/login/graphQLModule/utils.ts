import { QueryOptions } from '@apollo/client';
import { merge } from 'lodash';

import { GraphQLModuleModuleProps } from '.';
import { ApiTokenClient } from '../apiTokensClient';

type PartialContext = Partial<QueryOptions['context']>;
type QueryHeaders = Record<string, unknown>;

function mergeQueryOptionsContexts(target: PartialContext, overrides: PartialContext): PartialContext {
  return merge(target, overrides);
}

export function mergeQueryOptionContexts(options: QueryOptions, overrideContext: PartialContext): QueryOptions {
  const context: PartialContext = options.context || {};
  // eslint-disable-next-line no-param-reassign
  options.context = mergeQueryOptionsContexts(context, overrideContext);
  return options;
}

export function mergeHeadersToQueryOptions(options: QueryOptions, headers: QueryHeaders): QueryOptions {
  const context: PartialContext = {
    headers,
  };
  return mergeQueryOptionContexts(options, context);
}

export function mergeAuthorizationHeaderToQueryOptions(options: QueryOptions, authHeader: string): QueryOptions {
  const headers: QueryHeaders = {
    authorization: authHeader,
  };
  return mergeHeadersToQueryOptions(options, headers);
}

export function mergeQueryOptions(
  target: Partial<GraphQLModuleModuleProps['queryOptions']>,
  overrides: Partial<GraphQLModuleModuleProps['queryOptions']>,
) {
  return merge(target, overrides);
}

/* No need to export to hds-react package */
export function mergeQueryOptionsToModuleProps<T, Q>(
  target: Partial<GraphQLModuleModuleProps<T, Q>>,
  queryOptions: Partial<QueryOptions<Q>>,
) {
  const { query, ...rest } = queryOptions;
  // eslint-disable-next-line no-param-reassign
  target.queryOptions = mergeQueryOptions(target.queryOptions || {}, rest);
  if (query) {
    // eslint-disable-next-line no-param-reassign
    target.query = query;
  }
  return target;
}

export function setBearerToQueryOptions(queryOptions: QueryOptions, token: string) {
  if (!token) {
    return queryOptions;
  }
  return mergeAuthorizationHeaderToQueryOptions(queryOptions, `Bearer ${token}`);
}

export function appendFetchOptions(queryOptions: QueryOptions, newOptions: Partial<Parameters<typeof fetch>[1]>) {
  if (!queryOptions.context) {
    // eslint-disable-next-line no-param-reassign
    queryOptions.context = {};
  }

  // eslint-disable-next-line no-param-reassign
  queryOptions.context.fetchOptions = {
    ...queryOptions.context.fetchOptions,
    ...newOptions,
  };
  return queryOptions;
}

/* No need to export to hds-react package */
export function mergeQueryOptionModifiers({
  options,
  queryHelper,
}: Pick<GraphQLModuleModuleProps, 'queryHelper' | 'options'>): Required<GraphQLModuleModuleProps>['queryHelper'] {
  const { apiTokenKey } = options || {};
  if (!apiTokenKey && !queryHelper) {
    return (opt: QueryOptions) => opt;
  }

  if (!apiTokenKey && queryHelper) {
    return queryHelper;
  }

  const tokenPicker = (apiTokenClient: ApiTokenClient, key: string) => {
    if (!apiTokenClient) {
      return undefined;
    }
    const tokens = apiTokenClient.getTokens();
    return tokens ? tokens[key] : undefined;
  };

  const apiTokenSetter: GraphQLModuleModuleProps['queryHelper'] = (opt, apiTokenClient) => {
    if (!apiTokenKey || !apiTokenClient) {
      return opt;
    }
    const token = tokenPicker(apiTokenClient, apiTokenKey);
    if (token) {
      return setBearerToQueryOptions(opt, token);
    }
    return opt;
  };

  if (!queryHelper) {
    return apiTokenSetter;
  }

  return (opt, apiTokenClient, beacon) => {
    return queryHelper(apiTokenSetter(opt, apiTokenClient, beacon), apiTokenClient, beacon);
  };
}
