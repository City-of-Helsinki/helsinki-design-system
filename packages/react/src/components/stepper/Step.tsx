import React from 'react';

import '../../styles/base.module.css';
import styles from './Stepper.module.scss';
import { IconCheck, IconError, IconPlaybackPause } from '../../icons';
import classNames from '../../utils/classNames';
import { IconSize } from '../../icons/Icon.interface';

export enum StepState {
  available,
  completed,
  disabled,
  attention,
  paused,
}

export type StepProps = {
  /**
   * The label of the step
   */
  label: string;
  /**
   * The language of the step
   */
  language?: Language;
  /**
   * The index of the step
   */
  index: number;
  /**
   * A callback function for custom action on step click
   */
  onStepClick?: (event: React.MouseEvent<HTMLButtonElement>, stepIndex: number) => void;
  /**
   * A function for rendering a custom aria-label for step's state
   */
  renderCustomStateAriaLabel?: (stepIndex: number, state: StepState) => string;
  /**
   * A function for rendering a custom step count label
   */
  renderCustomStepCountLabel?: (stepIndex: number, totalNumberOfSteps: number) => string;
  /**
   * A boolean indicating whether step is in selected state
   */
  selected?: boolean;
  /**
   * A boolean indicating whether small variant is used
   */
  small?: boolean;
  /**
   * The state of the step
   */
  state: StepState;
  /**
   * The total number of steps
   */
  stepsTotal: number;
} & React.ComponentPropsWithoutRef<'button'>;

type Language = 'en' | 'fi' | 'sv' | string;

const getStepCountLabel = (language: Language, stepIndex: number, totalNumberOfSteps: number) => {
  return {
    en: `Step ${stepIndex + 1}/${totalNumberOfSteps}.`,
    fi: `Vaihe ${stepIndex + 1}/${totalNumberOfSteps}.`,
    sv: `Steg ${stepIndex + 1}/${totalNumberOfSteps}.`,
  }[language];
};

const states = {
  available: {
    fi: 'Valittavissa.',
    en: 'Available.',
    sv: 'Valbar.',
  },
  completed: {
    fi: 'Valmis.',
    en: 'Completed.',
    sv: 'Komplett.',
  },
  attention: {
    fi: 'Vaatii huomiota.',
    en: 'Needs attention.',
    sv: 'Behöver uppmärksamhet.',
  },
  paused: {
    fi: 'Keskeytetty.',
    en: 'Paused.',
    sv: 'Pausad.',
  },
};

const getStepState = (language: Language, state: StepProps['state']) => {
  if (state === StepState.disabled) {
    // Button has disabled attribute, which is enough for a11y
    return '';
  }
  return states[StepState[state]][language];
};

export const Step = React.forwardRef<HTMLButtonElement, StepProps>(
  (
    {
      label,
      language = 'fi',
      index,
      renderCustomStepCountLabel,
      small = false,
      state,
      selected,
      stepsTotal,
      renderCustomStateAriaLabel,
      onStepClick,
      ...rest
    }: StepProps,
    ref?: React.RefObject<HTMLButtonElement>,
  ) => {
    const composeAriaLabel = () => {
      const stepCountLabel = renderCustomStepCountLabel
        ? renderCustomStepCountLabel(index, stepsTotal)
        : getStepCountLabel(language, index, stepsTotal);

      let stateAriaLabel = renderCustomStateAriaLabel
        ? renderCustomStateAriaLabel(index, state)
        : getStepState(language, state);

      if (selected && state === StepState.available) {
        stateAriaLabel = '';
      }

      let labelWithPeriod = label;

      if (labelWithPeriod.slice(-1) !== '.') {
        labelWithPeriod += '.';
      }

      return [labelWithPeriod, stepCountLabel, stateAriaLabel].filter((lbl) => lbl).join(' ');
    };

    return (
      <div className={styles.stepContainer}>
        <button
          ref={ref}
          type="button"
          disabled={state === StepState.disabled}
          className={classNames(
            styles.step,
            selected && styles.selected,
            state === StepState.disabled && styles.disabled,
          )}
          aria-current={selected ? 'step' : false}
          aria-label={composeAriaLabel()}
          onClick={(e) => onStepClick && onStepClick(e, index)}
          {...rest}
        >
          <div className={styles.circleContainer}>
            {state === StepState.completed && !selected ? (
              <div className={styles.completedContainer}>
                <IconCheck className={styles.completedIcon} />
              </div>
            ) : (
              <div className={classNames(styles.circle)}>
                {state === StepState.attention && <IconError size={IconSize.ExtraSmall} />}
                {state === StepState.paused && <IconPlaybackPause size={IconSize.ExtraSmall} />}
                {(state === StepState.available ||
                  state === StepState.disabled ||
                  (state === StepState.completed && selected)) && <span className={styles.number}>{index + 1}</span>}
              </div>
            )}
          </div>
          {!small && <p className={styles.label}>{label}</p>}
        </button>
      </div>
    );
  },
);
