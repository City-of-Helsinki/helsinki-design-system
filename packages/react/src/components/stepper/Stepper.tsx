import React, { createRef, useEffect, useRef, useState } from 'react';

import '../../styles/base.module.css';
import styles from './Stepper.module.scss';
import { Step, StepState } from './Step';
import classNames from '../../utils/classNames';
import { IconAngleLeft, IconAngleRight } from '../../icons';
import { useTheme } from '../../hooks/useTheme';
import { IconSize } from '../../icons/Icon.interface';
import { AllElementPropsWithoutRef } from '../../utils/elementTypings';

type Language = 'en' | 'fi' | 'sv' | string;

export interface StepperCustomTheme {
  /**
   * A custom background color for stepper
   */
  '--hds-stepper-background-color'?: string;
  /**
   * A custom color for stepper
   */
  '--hds-stepper-color'?: string;
  /**
   * A custom width for step
   */
  '--hds-step-width'?: string;
  /**
   * A custom color for selected step label
   */
  '--hds-selected-step-label-color'?: string;
  /**
   * A custom color for not selected step label
   */
  '--hds-not-selected-step-label-color'?: string;
  /**
   * A custom color for step circle background
   */
  '--hds-step-background-color'?: string;
  /**
   * A custom color for step circle content
   */
  '--hds-step-content-color'?: string;
  /**
   * A custom color for stepper focus borders
   */
  '--hds-stepper-focus-border-color'?: string;
}

type Steps = {
  /**
   * The state of the step
   */
  state: StepState;
  /**
   * The label of the step
   */
  label: string;
}[];

export type StepperProps = AllElementPropsWithoutRef<'div'> & {
  /**
   * A custom className passed to stepper
   */
  className?: string;
  /**
   * A custom class name for step heading
   */
  headingClassName?: string;
  /**
   * The language of the stepper
   */
  language?: Language;
  /**
   * A callback function for custom action on step click
   */
  onStepClick?: (event: React.MouseEvent<HTMLButtonElement>, stepIndex: number) => void;
  /**
   * A function for rendering a custom aria-label for step's state
   */
  renderCustomStateAriaLabel?: (stepIndex: number, state: StepState) => string;
  /**
   * A function for rendering a custom step count label
   */
  renderCustomStepCountLabel?: (stepIndex: number, totalNumberOfSteps: number) => string;
  /**
   * A function for rendering a custom step heading
   */
  renderCustomStepHeading?: (stepIndex: number, totalNumberOfSteps: number, label: string) => string;
  /**
   * The index of the selected step. Used to set the aria-current="step" attribute to the active step
   */
  selectedStep?: number;
  /**
   * Boolean indicating small variant usage
   */
  small?: boolean;
  /**
   * A boolean indicating step heading variant usage
   */
  stepHeading?: boolean;
  /**
   * Step heading aria level
   */
  stepHeadingAriaLevel?: number;
  /**
   * The steps of the stepper
   */
  steps: Steps;
  /**
   * Custom theme
   */
  theme?: StepperCustomTheme; // Custom theme styles
};
const getStepHeading = (language: Language, stepIndex: number, totalNumberOfSteps: number, label: string) => {
  return {
    en: `Step ${stepIndex + 1}/${totalNumberOfSteps}: ${label}`,
    fi: `Vaihe ${stepIndex + 1}/${totalNumberOfSteps}: ${label}`,
    sv: `Steg ${stepIndex + 1}/${totalNumberOfSteps}: ${label}`,
  }[language];
};

export const Stepper = ({
  className,
  language = 'fi',
  selectedStep,
  small = false,
  renderCustomStepCountLabel,
  renderCustomStateAriaLabel,
  onStepClick,
  stepHeading,
  stepHeadingAriaLevel = 2,
  headingClassName,
  'data-testid': dataTestId,
  renderCustomStepHeading,
  steps,
  theme,
  ...rest
}: StepperProps) => {
  const stepsTotal = steps.length;
  const initialRender = useRef(true);
  const stepHeadingRef = useRef(null);
  const stepperRef = useRef(null);
  const stepRefs = useRef([]);
  const arrLength = steps.length;
  const [showPreviousButton, setShowPreviousButton] = useState(false);
  const [showNextButton, setShowNextButton] = useState(false);
  const customThemeClass = useTheme<StepperCustomTheme>(styles.stepperContainer, theme);

  if (stepRefs.current.length !== arrLength) {
    // add or remove refs
    stepRefs.current = Array(arrLength)
      .fill(0)
      .map((_, index) => stepRefs.current[index] || createRef());
  }

  useEffect(() => {
    if (!initialRender.current) {
      if (stepHeadingRef.current) {
        stepHeadingRef.current.focus();
      }

      const hdsStepWidth = stepperRef.current.scrollWidth / stepsTotal;

      stepperRef.current.scrollLeft =
        hdsStepWidth * (selectedStep + 1) - hdsStepWidth / 2.0 - stepperRef.current.parentNode.clientWidth / 2.0;
    }

    initialRender.current = false;

    if (stepperRef.current.scrollLeft > 5) {
      setShowPreviousButton(true);
    } else {
      setShowPreviousButton(false);
    }

    if (
      stepperRef.current.scrollWidth - (stepperRef.current.parentNode.clientWidth + stepperRef.current.scrollLeft) >
      5
    ) {
      setShowNextButton(true);
    } else {
      setShowNextButton(false);
    }
  }, [selectedStep]);

  return (
    <div
      {...rest}
      lang={language}
      className={classNames(styles.stepperContainer, customThemeClass)}
      data-testid={dataTestId}
    >
      {showPreviousButton && (
        <div className={classNames(styles.scrollButton, styles.scrollButtonPrevious)} aria-hidden="true">
          {/*  eslint-disable-next-line jsx-a11y/control-has-associated-label */}
          <button
            type="button"
            onClick={() => {
              const amountToScroll = stepperRef.current.scrollWidth / stepsTotal;
              stepperRef.current.scrollLeft -= amountToScroll;
            }}
            tabIndex={-1}
          >
            <IconAngleLeft size={IconSize.Medium} />
          </button>
        </div>
      )}
      {showNextButton && (
        <div className={classNames(styles.scrollButton, styles.scrollButtonNext)} aria-hidden="true">
          {/*  eslint-disable-next-line jsx-a11y/control-has-associated-label */}
          <button
            type="button"
            onClick={() => {
              const amountToScroll = stepperRef.current.scrollWidth / stepsTotal;
              stepperRef.current.scrollLeft += amountToScroll;
            }}
            tabIndex={-1}
          >
            <IconAngleRight size={IconSize.Medium} />
          </button>
        </div>
      )}
      <div
        onScroll={(e) => {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          if (e.target.scrollLeft && e.target.scrollLeft > 5) {
            setShowPreviousButton(true);
          } else {
            setShowPreviousButton(false);
          }
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          if (e.target.scrollWidth - (e.target.parentNode.clientWidth + e.target.scrollLeft) > 5) {
            setShowNextButton(true);
          } else {
            setShowNextButton(false);
          }
        }}
        ref={stepperRef}
        className={classNames(className, styles.stepper, small && styles.small)}
      >
        <div
          className={styles.line}
          aria-hidden
          style={{
            width: `max( calc(100% - var(--hds-step-width)), calc(${stepsTotal} * var(--hds-step-width) - var(--hds-step-width) ))`,
          }}
        >
          {steps.map((step, index) => {
            if (index === steps.length - 1) {
              return null;
            }
            return (
              <div
                key={`${step.label}-${index}`} // eslint-disable-line react/no-array-index-key
                style={{ width: `calc( 100% / ${stepsTotal - 1})` }}
                className={steps[index + 1].state === StepState.disabled ? styles.disabledLine : styles.enabledLine}
              />
            );
          })}
        </div>
        {steps.map((step, index) => {
          return (
            <Step
              ref={stepRefs.current[index]}
              key={`${index}-${step.label}`} // eslint-disable-line react/no-array-index-key
              label={step.label}
              language={language}
              index={index}
              small={small}
              stepsTotal={stepsTotal}
              selected={selectedStep === index}
              state={step.state}
              onStepClick={(event, stepIndex) => {
                if (stepIndex === selectedStep && stepHeading) {
                  stepHeadingRef.current.focus();
                }
                onStepClick(event, stepIndex);
              }}
              renderCustomStepCountLabel={renderCustomStepCountLabel}
              renderCustomStateAriaLabel={renderCustomStateAriaLabel}
              data-testid={dataTestId ? `${dataTestId}-step-${index}` : undefined}
            />
          );
        })}
      </div>
      {stepHeading && (
        <div
          tabIndex={-1}
          ref={stepHeadingRef}
          role="heading"
          aria-level={stepHeadingAriaLevel}
          className={classNames(styles.heading, headingClassName)}
        >
          {renderCustomStepHeading
            ? renderCustomStepHeading(selectedStep, stepsTotal, steps[selectedStep].label)
            : getStepHeading(language, selectedStep, stepsTotal, steps[selectedStep].label)}
        </div>
      )}
    </div>
  );
};
