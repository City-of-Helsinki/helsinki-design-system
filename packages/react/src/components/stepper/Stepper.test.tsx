/* eslint-disable react/forbid-component-props */
import React, { HTMLAttributes } from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { Stepper } from './Stepper';
import { StepState } from './Step';
import { getCommonElementTestProps, getElementAttributesMisMatches } from '../../utils/testHelpers';

describe('<Stepper /> spec', () => {
  let state;

  beforeEach(() => {
    // Need to mock scrollIntoView for jest
    window.HTMLElement.prototype.scrollIntoView = function () {}; // eslint-disable-line
    state = {
      activeStepIndex: 0,
      steps: [
        {
          state: StepState.available,
          label: 'Step 1',
        },
        {
          state: StepState.disabled,
          label: 'Step 2',
        },
        {
          state: StepState.disabled,
          label: 'Step 3',
        },
        {
          state: StepState.disabled,
          label: 'Step 4 - longer text',
        },
        {
          state: StepState.disabled,
          label: 'Step 5',
        },
      ],
    };
  });

  it('renders the component', () => {
    const { asFragment } = render(
      <Stepper steps={state.steps} language="en" selectedStep={state.activeStepIndex} data-testid="hds-stepper" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should be able to select available step', () => {
    const { container, rerender } = render(
      <Stepper steps={state.steps} language="en" selectedStep={state.activeStepIndex} data-testid="hds-stepper" />,
    );
    const availableButton = container.querySelector('[data-testid="hds-stepper-step-1"]');
    expect(availableButton).toHaveAttribute('aria-current', 'false');

    rerender(
      <Stepper
        steps={[
          {
            state: StepState.available,
            label: 'Step 1',
          },
          {
            state: StepState.available,
            label: 'Step 2',
          },
          {
            state: StepState.disabled,
            label: 'Step 3',
          },
          {
            state: StepState.disabled,
            label: 'Step 4 - longer text',
          },
          {
            state: StepState.disabled,
            label: 'Step 5',
          },
        ]}
        language="en"
        selectedStep={1}
        data-testid="hds-stepper"
      />,
    );
    expect(availableButton).toHaveAttribute('aria-current', 'step');
  });

  it('should not have basic accessibility issues', async () => {
    const { container } = render(
      <Stepper language="en" steps={state.steps} selectedStep={state.activeStepIndex} data-testid="hds-stepper" />,
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
  it('native html props are passed to the element', async () => {
    const divProps = getCommonElementTestProps<'div'>('div');
    const { getByTestId } = render(
      <Stepper language="en" steps={state.steps} selectedStep={state.activeStepIndex} {...divProps} />,
    );
    const element = getByTestId(divProps['data-testid']);
    // className is applied to another element
    expect(
      getElementAttributesMisMatches(element, {
        ...divProps,
        className: undefined,
      } as HTMLAttributes<HTMLDivElement>),
    ).toHaveLength(0);
  });
});
