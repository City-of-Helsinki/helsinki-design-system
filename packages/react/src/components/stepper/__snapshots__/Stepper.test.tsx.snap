// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stepper /> spec renders the component 1`] = `
<DocumentFragment>
  <div
    class="stepperContainer"
    data-testid="hds-stepper"
    lang="en"
  >
    <div
      class="stepper"
      tabindex="0"
    >
      <div
        aria-hidden="true"
        class="line"
      >
        <div
          class="disabledLine"
          style="width: calc( 100% / 4);"
        />
        <div
          class="disabledLine"
          style="width: calc( 100% / 4);"
        />
        <div
          class="disabledLine"
          style="width: calc( 100% / 4);"
        />
        <div
          class="disabledLine"
          style="width: calc( 100% / 4);"
        />
      </div>
      <div
        class="stepContainer"
      >
        <button
          aria-current="step"
          aria-label="Step 1. Step 1/5. Selected."
          class="step selected"
          data-testid="hds-stepper-step-0"
          tabindex="-1"
          type="button"
        >
          <div
            class="circleContainer"
          >
            <div
              class="circle"
            >
              <span
                class="number"
              >
                1
              </span>
            </div>
          </div>
          <p
            class="label"
          >
            Step 1
          </p>
        </button>
      </div>
      <div
        class="stepContainer"
      >
        <button
          aria-current="false"
          aria-label="Step 2. Step 2/5. Disabled."
          class="step disabled"
          data-testid="hds-stepper-step-1"
          disabled=""
          tabindex="0"
          type="button"
        >
          <div
            class="circleContainer"
          >
            <div
              class="circle"
            >
              <span
                class="number"
              >
                2
              </span>
            </div>
          </div>
          <p
            class="label"
          >
            Step 2
          </p>
        </button>
      </div>
      <div
        class="stepContainer"
      >
        <button
          aria-current="false"
          aria-label="Step 3. Step 3/5. Disabled."
          class="step disabled"
          data-testid="hds-stepper-step-2"
          disabled=""
          tabindex="0"
          type="button"
        >
          <div
            class="circleContainer"
          >
            <div
              class="circle"
            >
              <span
                class="number"
              >
                3
              </span>
            </div>
          </div>
          <p
            class="label"
          >
            Step 3
          </p>
        </button>
      </div>
      <div
        class="stepContainer"
      >
        <button
          aria-current="false"
          aria-label="Step 4 - longer text. Step 4/5. Disabled."
          class="step disabled"
          data-testid="hds-stepper-step-3"
          disabled=""
          tabindex="0"
          type="button"
        >
          <div
            class="circleContainer"
          >
            <div
              class="circle"
            >
              <span
                class="number"
              >
                4
              </span>
            </div>
          </div>
          <p
            class="label"
          >
            Step 4 - longer text
          </p>
        </button>
      </div>
      <div
        class="stepContainer"
      >
        <button
          aria-current="false"
          aria-label="Step 5. Step 5/5. Disabled."
          class="step disabled"
          data-testid="hds-stepper-step-4"
          disabled=""
          tabindex="0"
          type="button"
        >
          <div
            class="circleContainer"
          >
            <div
              class="circle"
            >
              <span
                class="number"
              >
                5
              </span>
            </div>
          </div>
          <p
            class="label"
          >
            Step 5
          </p>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
